### PR TITLE
Add type annotations

### LINKED ISSUE #421: Add type annotations
Type annotations are helpful both to catch errors and as extra documentation for developers trying to make sense of the code.

### PR DESCRIPTION
**Pull request by mthuurne**
# Description
Adds type annotations and checks them using mypy in strict mode.

I omitted the unit tests, as they use Trial's `assertSomething()` methods which aren't annotated yet in Twisted, so there would not be much value in annotating the tests at this time.

Closes #421

# Checklist

* [X] Make sure changes are covered by existing or new tests. (mypy checks the annotations)
* [X] For at least one Python version, make sure local test run is green. (py38, py39 and py310 pass)
* [X] Create a file in `src/towncrier/newsfragments/`. Describe your
  change and include important information. Your change will be included in the public release notes.  (left empty because it's not relevant to end users)
* [X] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).

### CODE REVIEW

File: src/towncrier/_project.py
Comment (mthuurne) [inline src/towncrier/_project.py:83]: The previous code would implicitly return `None` here, but the caller couldn't handler `None`, so I believe that was a bug.

File: src/towncrier/_settings/load.py
Comment (adiroiban) [inline src/towncrier/_settings/load.py]: I guess that this is private API so is ok to rename the argument.

**Review by adiroiban — APPROVED** [review]
Looks good to me, but I have little experience with mypy.

I have enabled mypy on CI to make sure we continue to keep annotation in future changes.

Also the new tests looks ok.

**Comment by adiroiban** [comment]
Many thanks @mthuurne for the update.

I did a quick review.
I added back the team.

You can wait for a review or feel free to merge this.

Thanks again

File: src/towncrier/_settings/load.py
Comment (mthuurne) [reply src/towncrier/_settings/load.py]: Mypy doesn't like one variable being used with two different types, so I had to rename either the argument or the local use. As the argument contains the path of the config file rather than the configuration itself, I figured that would be the best one to rename. And as you say, it's inside a private module, so it should be safe.

By the way, does Towncrier even have a public API in the form of Python modules or is the command line interface the only public interface? The docs don't mention any modules and the docstrings are rather sparse.

**Comment by mthuurne** [comment]
I'll wait a day or so for any other devs to give their opinion and merge if I get an additional approval or no response.

**Comment by hynek** [comment]
I'll try to find more time to look at it, but two things for now:

- put the config into pyproject.toml
- add `from __future__ import annotations` to all files with type stubs and run Pyupgrade (part of pre-commit)

File: MANIFEST.in
Comment (hynek) [inline MANIFEST.in]:
```suggestion
```

File: src/towncrier/_writer.py
Comment (hynek) [inline src/towncrier/_writer.py]: Please don't re-add defaults to private APIs. 😅

```suggestion
    single_file: bool,
```

File: src/towncrier/create.py
Comment (hynek) [inline src/towncrier/create.py:68]: I'm confused by both the change and why mypy isn't complaining, even if I add `warn_no_return = true`. What is happening her? Did they change their stance on implicit None returns?

**Review by hynek — CHANGES_REQUESTED** [review]
Looks mostly good, should be ready for merge after a few nits!

File: src/towncrier/_writer.py
Comment (mthuurne) [reply src/towncrier/_writer.py]: Ah, that one slipped through when I rebased my changes on `trunk`.

File: src/towncrier/create.py
Comment (mthuurne) [reply src/towncrier/create.py:68]: And now I'm confused by your confusion. ;) Why would mypy complain about it? And would you expect it to complain about the old code, the new code or both?

What happened is that when annotating `_main()` (single), I looked at what values `__main()` (double) can return and saw that it doesn't have a `return` statement. Therefore I thought it was cleaner to not return the implicit `None` but instead remove the `return` from `_main()` (single).

Comment (mthuurne) [reply src/towncrier/create.py:68]: `warn_no_return` is documented as "Shows errors for missing return statements on some execution paths.", but `__main()` (double) has no `return` statement on **any** execution path.

Comment (hynek) [reply src/towncrier/create.py:68]: ugh yeah you're entirely correct

**Review by hynek — APPROVED** [review]
Thanks for this!. Types are a huge boon to a low-churn project like towncrier and will help me with some refactorings I have in mind.

**Pull request merged**
